### PR TITLE
Fix #2203

### LINKED ISSUE #2203: Custom types with same name in different schemas don't work due to oid caching
When you have a custom type (e.g. using `diesel-derive-enum`) with the same name in different schemas, it doesn't work with diesel, it doesn't consider the search_path when looking up types.
Instead, diesel uses the first type that matches the name, regardless of the schema it's in.

It seems like this PR introduced this behavior: https://github.com/diesel-rs/diesel/pull/2088

Btw, it looks like jdbc had the same issue:
https://www.postgresql.org/message-id/flat/20110308183723.08A5B1337B93%40mail.postgresql.org

### PR DESCRIPTION
**Pull request by weiznich**
Custom postgres types could live in different schemas, so we need to
incorporate the corresponding schema in the lookup request. This commit
adds an optional parameter to `#[derive(SqlType)]` that allows to
configure the corresponding schema. Otherwise the default search path is used.

### CODE REVIEW

**Review by pksunkara — APPROVED** [review]

**Comment by pksunkara** [comment]
The proceess is to run `TRYBUILD=overwrite cargo test` locally to overwrite the compile outputs and then check the diff and commit them.

File: diesel_compile_tests/tests/fail/array_expressions_must_be_same_type.stderr [hunk 1]
```diff
@@ -106,7 +106,7 @@ error[E0277]: the trait bound `{integer}: diesel::SelectableExpression<()>` is n
               <(A, B) as diesel::SelectableExpression<QS>>
               <(A, B, C) as diesel::SelectableExpression<QS>>
               <(A, B, C, D) as diesel::SelectableExpression<QS>>
-            and 108 others
```
Comment (weiznich) [inline diesel_compile_tests/tests/fail/array_expressions_must_be_same_type.stderr:109]: @pksunkara It's not clear for me why this change is happening here. The old `compile_tests` setup didn't have this "problem".

File: diesel/src/pg/metadata_lookup.rs [hunk 2]
```diff
@@ -24,26 +25,87 @@ impl PgMetadataLookup {
...
+
+        metadata_cache.lookup_type(&cache_key).unwrap_or_else(|| {
+            let r = lookup_type(&cache_key, &self.conn);
+
+            match r {
+                Ok(type_metadata) => {
+                    metadata_cache.store_type(cache_key, type_metadata);
+                    type_metadata
+                }
+                Err(_e) => {
```
[the remark this answers is not in the document]
Comment (weiznich) [reply diesel/src/pg/metadata_lookup.rs:43]: That's a good discussion point. There is no 1.4.x with metadata caching, therefore the current default behaviour is to just run the query every time we need the type oid. Another problem with the old implementation is that it does just silently accept that a type is non-existent and just returned 0 in those cases. Now this works most of the time but lead to surprising bugs in the past (Something like this works everywhere, but for this one query that requires something to be bound as array element).
The question is how to correctly handle cases where the oid does not exist (yet?):
* Just `panic!` , which I believe is a bad solution
* Change `HasSqlType::metadata` to return a result and return an error in such cases? That's a quite large change, not sure if that's worth it. (+ other database systems cannot fail there)
* Log and error and do that what should work most of the time. That's this solution. Now I've gone for the variant that requires that types anytime they are requested, as the type oid could appear, because the type is created dynamically (thought that's quite unlikely).

In general postgres system of allowing to change/add a new type at runtime is quite hard to come by with caching. There will likely always be a case that slip through and breaks things. The only bullet proof variant seems to be to fetch everything each request, but that's obviously bad for performance.

File: diesel/src/pg/metadata_lookup.rs [hunk 3]
```diff
@@ -24,26 +25,87 @@ impl PgMetadataLookup {
...
+fn lookup_type(
+    cache_key: &PgMetadataCacheKey<'_>,
+    conn: &PgConnection,
+) -> QueryResult<PgTypeMetadata> {
+    let search_path: String;
+
+    let search_schema = if let Some(schema) = cache_key.schema.as_ref() {
+        vec![schema.as_ref()]
+    } else {
+        search_path = crate::dsl::sql("SHOW search_path").get_result::<String>(conn)?;
```
[the remark this answers is not in the document]
Comment (Ten0) [reply diesel/src/pg/metadata_lookup.rs:67]: Hmm so that's a bugfix then. :+1:

File: diesel/src/pg/metadata_lookup.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ten0) [reply diesel/src/pg/metadata_lookup.rs:43]: > The only bullet proof variant seems to be to fetch everything each request, but that's obviously bad for performance.

[I initially implemented this caching](https://github.com/diesel-rs/diesel/pull/2088) with your help precisely because the performance impact turned out to be extremely high ^^

> I've gone for the variant that requires that types anytime they are requested, as the type oid could appear, because the type is created dynamically (thought that's quite unlikely).

The not-so-unlikely scenario that would be an issue for us is:
 - OID is not found because it's badly declared (e.g. it's in the wrong schema, or it has the wrong name due to a typo)
 - Postgres silent-recovers from the OID not being found and infers it all the time
 - The query costs every time in case it would be found, sliently (or almost, possibly generating lots of logs) having a huge perf impact

So I believe we have to make a choice between:
1. Letting PG attempt to recover without giving it the type OID
2. Re-sending query if OID was not found last time (in case it's created dynamically *after a previous fail*)

Indeed, not doing 1 would prevent the silent recover (forcing a fix -> no silent perf impact), and not doing 2 would prevent re-sending the query (-> no perf impact).

Now as I don't like silent fails I'd favor dropping 1 but it looks like dropping 1 would, as you mentioned, require either of:

> * Just `panic!` , which I believe is a bad solution
> * Change `HasSqlType::metadata` to return a result and return an error in such cases? That's a quite large change, not sure if that's worth it. (+ other database systems cannot fail there)

I don't like panicking in this scenario either, so if changing `HasSqlType::metadata` to return a result is too complex then it looks like we should drop 2 instead.

[the remark this answers is not in the document]
Comment (weiznich) [reply diesel/src/pg/metadata_lookup.rs:44]: I've gone forward to make this actually return an error if such a unknown type is used somewhere, by wrapping this information as part of the `PgTypeMetadata` struct. This somehow seems to be the correct solution for this problem, even if that's the larger change. As part of this change I've made that struct opaque as this could potentially help with #2087 later.

Another thing we could do here is to use case insensitive comparisons for the type and schema names as postgres does not care about this internally. I think all type names are lower cased internally or something like that.

[the remark this answers is not in the document]
Comment (Ten0) [reply diesel/src/pg/metadata_lookup.rs:44]: So I haven't had time to look at every usage of TypeMetadata yet so this may be imprecise, but with the error possibly wrapped in `PgTypeMetadata` and only triggering when attempting to read the inner values, I'm not sure I see what guarantees there isn't a scenario where we build `PgTypeMetadata`, possibly resulting in wrapping an error inside (not storing cache), then actually not use it -> not triggering the error. If there was such a scenario, then my scenario at https://github.com/diesel-rs/diesel/pull/2628/files#r564001970 could still happen.

What guarantees that actually attempt to extract an OID out of `PgTypeMetadata` everytime we've built ?
If that is hard to guarantee, maybe we want to make `lookup_type` directly return a `Result`, and `?` it right away everytime ?

[the remark this answers is not in the document]
Comment (weiznich) [reply diesel/src/pg/metadata_lookup.rs:44]: True, it's possible to just call `SomeSqlType::metadata(lookup)` somewhere without using the result. For diesel itself I would consider that as a bug (and I'm really sure we are using it everywhere), for third party code that's up to whatever they want to do there.
I've added at `#[must_use]` to `PgTypeMetadata` to make it at least warn if we would construct an instance of that type without using it. 

I've opted against making `HasSqlType::metadata()` return as this is a much larger change and all of the other backends have a closed not user extensible set of type metadata.

[the remark this answers is not in the document]
Comment (Ten0) [reply diesel/src/pg/metadata_lookup.rs:44]: > I've added at `#[must_use]` to `PgTypeMetadata`
> I'm really sure we are using it everywhere

Alright :)

**Review by Ten0 — APPROVED** [review]
There's still the missing backquote, otherwise LGTM :)
